docs/proposals/vault-auth-provider.rst

```diff
@@ -0,0 +1,154 @@
+
+Background
+==========
+
+Glossary
+--------
+
+-  IAM Role (iam-role) - an IAM principal. Our instances have an IAM
+   role attached which they use as a service account.
+   https://docs.aws.amazon.com/IAM/latest/UserGuide/id\_roles.html
+-  IAM Policy (iam-policy) - the authz rules (and rules documents) in
+   AWS.
+-  Vault Token (vault-token) - common identity document given by all
+   vault auth providers. Has a list of policy (documents).
+-  Vault Policy (vault-policy) - both the policy language itself (giving
+   authorization for vault operations), document of those policy
+   statement,s and the list of such documents named on a vault-token.
+   vault-policy has its own type, which we'll call *policy*
+-  Vault PKI Role (pki-role) - vault-policy isn't very powerful wrt to
+   PKI operations, e.g. it doesn't let you lock PKI certificate issuance
+   down to specific CNs etc. The PKI secrets provider has its own
+   mechanism for this; the PKI role. pki-roles live at paths and access
+   is controlled by vault-policy. pki-roles are implemented as vault
+   generic secrets.
+   https://www.vaultproject.io/api/secret/pki/index.html#create-update-role
+
+Problem
+-------
+
+We would like to have individual Vault policies per EC2 instance, so
+that they can't issue themselves certificates for anyone else's FQDN.
+
+See:
+
+- https://github.com/jetstack/tarmak/issues/120
+- https://github.com/jetstack/tarmak/issues/34
+
+Posibilities
+------------
+
+Due to several Vault limitations, especially in the free version, it's
+not possible to:
+
+- Add a vault-policy (name) to a token after it's been created (by vault ec2 login)
+- Specify a vault-policy expressive enough to re-use one policy for all EC2 Instances (i.e. no variable back-references)
+- Vault Premium has a different policy language called Sentinel, which looks like it could do this
+- Use the k8s auth provider, as we also want to use this mechanism with etcd Instances, etc, which do not run Kubelet
+
+Objective
+=========
+
+Solution
+--------
+
+Our proposed solution is Vault auth provider. This will auth EC2 instances using their IID like the current AWS auth provider does, but it will also generate new vault-policies and pki-roles on the fly, which will e.g. limit their cert creation power to just that instance's DNS name.
+
+This vault auth plugin will serve exactly one environment (i.e. all kubernetes clusters in the same (provider, region))
+
+On login, our provider will:
+
+- Auth the Instance like the AWS provider's EC2 mode does (can we simply defer to that code?)
+- Match the iam-role attached to the Instance against our Provider's config, using its ARN
+- Make a pki-role from the configured template for that iam-role
+- Make a vault-policy (document) from the configured template for that iam-role, including templating in the name of the pki-role
+
+  - This should have a unique name based on the AWS Instance ID / boot ID / etc
```

Comment:
As an alternative, could something in the `tools` stack respond to new instances being created, and automatically provision the roles etc. ready for the instance? & add some retry/check logic in wing to ensure the correct policy/permissions are associated with the token received?

Reply:
Perhaps `wing-controller` (not sure if this exists yet) could watch for `Wing` resources in the apiserver, and auto provision the appropriate node roles in Vault? This would save the dependance on the single tools stack for all clusters (which should make portability between clouds, where they may be just one tools stack in one of the clouds (not sure if this'll be the case, but just in case it is))

Reply:
Yeah right now wing is not involved, if we scale up a prebacked worker image, before the node becomes healthy and usuable.

Making wing a vault controller, would require a (more secure and HA) setup of wing. Then we would need a tarmak for setting up wing :see_no_evil: 

Giving wing that unneeded

Reply:
I'd have thought a secure deployment of Wing is a goal anyway though?

HA is a consideration, however if Wing were to go unavailable, all existing instances (i.e. ones that have had policies provisioned for them already), would continue to work, and should also be able to renew their tokens just fine.

This means a Wing outage = no *new* instances can be created, but existing ones can reboot etc.

Reply:
Yeah but currently wing is not needed for new instances to come up, it is just an orchestratior for updates and reporting back the instance status.

Not being able to bringing up new instances would make a huge requirement for setting up wing in a HA manner. All other components involved in the bootstrap are HA as of today.

Reply:
Sorry, thought I'd written about that: another major concern is that this would be a race. After we know an instance is booting (not necessarily under our control if it's say a recovery), then we have to get the role ready /before/ that instance tries to auth with Vault, as the list of roles in a token is static.